### PR TITLE
[1858] Stock round: auctions, buying and selling shares

### PR DESCRIPTION
**Pull request by ollybh**
This is the start of the stock round implementation for 1858. Buying and selling shares is fairly normal, the main difference being that you can't directly par a corporation before phase 5. Before that point corporations can only be started by exchanging a private railway company, this is to be added in a later commit.

In a stock round you can also put up a private company for auction. Everyone gets a chance to bid on this company, until only one person is left. This is implemented using the Engine::Step::BuySellParSharesViaBid code from 1861/1867.

### CODE REVIEW

File: lib/engine/game/g_1858/step/buy_sell_par_shares.rb [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+          end
+
+          def can_bid_company?(player, company)
+            (@auctioning.nil? || @auctioning == company) &&
+            auctionable_companies.include?(company) &&
+              (min_bid(company) <= player.cash)
+          end
+
+          def log_skip(entity)
+            super unless @round.current_actions.any?
```
Comment (tobymao) [inline lib/engine/game/g_1858/step/buy_sell_par_shares.rb:108]: if empty?

File: lib/engine/game/g_1858/step/buy_sell_par_shares.rb [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+          def auctionable_companies
+            @game.buyable_bank_owned_companies
+          end
+
+          def can_bid_any?(player)
+            auctionable_companies.any? { |company| can_bid_company?(player, company) }
+          end
+
+          def can_bid_company?(player, company)
+            (@auctioning.nil? || @auctioning == company) &&
```
Comment (tobymao) [inline lib/engine/game/g_1858/step/buy_sell_par_shares.rb:102]: !@auctioning

File: lib/engine/game/g_1858/step/buy_sell_par_shares.rb (hunk 1 above)
Comment (ollybh) [reply lib/engine/game/g_1858/step/buy_sell_par_shares.rb]: Sorry, I thought I'd already got rid of all of these blockless `any?` and `none?` calls. It's gone now.

File: lib/engine/game/g_1858/step/buy_sell_par_shares.rb (hunk 2 above)
Comment (ollybh) [reply lib/engine/game/g_1858/step/buy_sell_par_shares.rb]: Changed.

**Pull request merged**